File: src/main/java/dev/rvbsm/fsit/api/event/ClientCommandCallback.java
package dev.rvbsm.fsit.api.event;

import net.minecraft.network.packet.c2s.play.ClientCommandC2SPacket;
import net.minecraft.server.network.ServerPlayerEntity;
import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;

import org.jetbrains.annotations.NotNull;

@FunctionalInterface
public interface ClientCommandCallback {

    Event<ClientCommandCallback> EVENT = EventFactory.createArrayBacked(
        ClientCommandCallback.class, (listeners) -> (player, mode) -> {
            for (ClientCommandCallback listener : listeners) {
                listener.process(player, mode);
            }
        });

    void process(@NotNull ServerPlayerEntity player, @NotNull ClientCommandC2SPacket.Mode mode);
}

File: src/main/kotlin/dev/rvbsm/fsit/event/ClientCommandCallback.kt
package dev.rvbsm.fsit.event

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.network.packet.c2s.play.ClientCommandC2SPacket
import net.minecraft.server.network.ServerPlayerEntity

fun interface ClientCommandCallback {
    fun onClientMode(player: ServerPlayerEntity, mode: ClientCommandC2SPacket.Mode)

    companion object {
        @JvmField
        val EVENT: Event<ClientCommandCallback> =
            EventFactory.createArrayBacked(ClientCommandCallback::class.java) { listeners ->
                ClientCommandCallback { player, mode ->
                    for (listener in listeners) {
                        listener.onClientMode(player, mode)
                    }
                }
            }
    }
}
